### ISSUE TITLE
Architecture question

### ISSUE THREAD

**Issue by hungyanbin** [bufferapp/android-clean-architecture-boilerplate#28]
According to this talk : https://www.youtube.com/watch?v=Nsjsiz2A9mg
Uncle bob says that a project's structure should not be MVC or MVP, MVVM...,
But this project top structure is mobile-ui/presentation/domain/data...,
Programmers cannot know what is this app doing at first glimpse, could you tell me what do you think about this? I am so confused. Thank you!!

**Comment by ataulm** [comment]
@hungyanbin Bob Martin wrote about the [dependency inversion principle (PDF)](https://web.archive.org/web/20110714224327/http://www.objectmentor.com/resources/articles/dip.pdf) - it's a long read, but worth it to understand the concept behind clean architecture.

This project demonstrates clean architecture for Android apps. To understand the structure, it's helpful to draw a dependency graph.

We can see from the `settings.gradle` that there are six modules:

- `cache`
- `remote`

**Comment by ataulm** [comment]
@hungyanbin Bob Martin wrote about the [dependency inversion principle (PDF)](https://web.archive.org/web/20110714224327/http://www.objectmentor.com/resources/articles/dip.pdf) - it's a long read, but worth it to understand the concept behind clean architecture.

This project demonstrates clean architecture for Android apps. To understand the structure, it's helpful to draw a dependency graph.

We can see from the `settings.gradle` that there are six modules:

```
include ':cache'
include ':remote'
include ':data'
include ':domain'
include ':presentation'
include ':mobile-ui'
```

Since this is an Android application - look for the module where the `build.gradle` applies the `com.android.application` plugin. It's `mobile-ui`.

From there, we can understand how the modules depend on each other (draw our graph).

We see that `mobile-ui` depends on four modules:

```
implementation project(':presentation')
implementation project(':data')
implementation project(':cache')
implementation project(':remote')
```

- `cache` depends on `data`
- `remote` depends on `data`
- `data` depends on `domain`
- `presentation` depends on `domain`

![img_20180420_222213](https://user-images.githubusercontent.com/2678555/39074496-9101ec2c-44e9-11e8-88f6-feb09e2cd33a.jpg)

(diagram misses the dependency arrow for `mobile-ui` on `data`)

Recall the diagram that I assume (😅) was in the video you linked (you can [see it on the 8th Light blog](https://8thlight.com/blog/uncle-bob/2012/08/13/the-clean-architecture.html) also).

![](https://user-images.githubusercontent.com/2678555/39072764-6eaa9774-44e3-11e8-8dba-60dd50aba128.png)

A simpler (more relevant) version [can be found here](https://github.com/android10/Android-CleanArchitecture).

![](https://user-images.githubusercontent.com/2678555/39072873-d1b38434-44e3-11e8-92c1-d3eb94e3b33f.png)

We could draw the module dependencies in a similar way, with:

- `domain` in the innermost circle
- then `presentation` and `data` which depend on `domain`
- then `cache` and `remote` which depend on `data`
- then `mobile-ui` which depends on `presentation`, `data`, `cache` and `remote`

Martin writes about _The Dependency Rule_:

>The concentric circles represent different areas of software. In general, the further in you go, the higher level the software becomes. The outer circles are mechanisms. The inner circles are policies.
>
>The overriding rule that makes this architecture work is The Dependency Rule. This rule says that source code dependencies can only point inwards. Nothing in an inner circle can know anything at all about something in an outer circle. In particular, the name of something declared in an outer circle must not be mentioned by the code in an inner circle. That includes, functions, classes. variables, or any other named software entity.
>
>By the same token, data formats used in an outer circle should not be used by an inner circle, especially if those formats are generate by a framework in an outer circle. We don’t want anything in an outer circle to impact the inner circles.

Also:

>The outermost layer is generally composed of frameworks and tools such as the Database, the Web Framework, etc. Generally you don’t write much code in this layer other than glue code that communicates to the next circle inwards.

That's what the `mobile-ui` module is in our case - the glue.

>Programmers cannot know what is this app doing at first glimpse

With an unfamiliar Android app, I find it easiest to approach by looking at the `AndroidManifest` - there we can see the entry point to the app: the `Activity` with the `android.intent.action.MAIN` intent filter, `BrowseActivity`. You can see that the dependencies are injected using Dagger in `onCreate()`, and then `onStart()`, the reins are handed over to the Presenter.

Where does the presenter live? In one of the inner modules, `presentation`. The presenter doesn't know about Android at all, and this is the idea - the closer you get to the inner circles, the closer you are to the logic/domain of your application, away from the boundaries of your system.

Did that help or did I ramble unnecessarily, not understanding your question? 🤔

**Comment by hungyanbin** [comment]
Thank you for your detailed explanation, you are right about Dependency rule. After these months, I bought the book "Clean Architecture", reading more materials. I found that there is a chapter :  The Missing chapter. That's exactly what I want !

Now I can ask the question more clearly - how to "package" the project? Should I package by layer(like this project)? package by feature(like pizza 🍕)? Or according to the book, package by module?

It is hard to decide which one is better, by layer or feature. Both dimensions are important, if I package by layer, the purpose of this APP is not clear(Chat app? Calendar app?). If I package by feature, dependency rule is easy to break. And package by module? The 'module' is responsibility is not easy to define.

For now, I choose package by module, but I still want to know other's opinion. And thank you again for your answer! It must takes long time to write.

**Comment by ataulm** [comment]
OH I see what you mean now and my answer is irrelevant :sweat_smile: :sweat_smile: :sweat_smile: :sweat_smile: 

I know @hitherejoe wrote about this topic:

- by feature - https://overflow.buffer.com/2016/09/26/android-rethinking-package-structure/
- new approach when moving to modularised project - https://overflow.buffer.com/2017/10/13/package-feature-modularised-android-projects/

~~I've not read the second one yet but seems like it'll be what you're after.~~

tldr: you can still package by feature, and then subpackages inside each feature for layer/component-based packages.

This project is probably too small (there's only the one screen/feature I think) to realise any meaningful feature-based packaging but [there are elements of it](https://github.com/bufferapp/android-clean-architecture-boilerplate/tree/master/presentation/src/main/java/org/buffer/android/boilerplate/presentation).

**Comment by hitherejoe** [comment]
@ataulm smashed it with that response, thanks for taking the time there 💪You're right about the mobile-ui module being the 'glue', it would be better if it was only aware of the presentation module in this kind of setup but the 'cache' and 'remote' modules are also within the Frameworks & Drivers layer so it's not so bad.

@hungyanbin as @ataulm has provided, there are some links there for packaging. This project is too small really for packaging by feature (I can change it if people would find it easier to follow). But for projects with more than one feature it would probably make more sense to follow something similar to [this](https://overflow.buffer.com/2017/10/13/package-feature-modularised-android-projects/).

I'd be happy to help out with anything if further clarification is needed :)

**Comment by hungyanbin** [comment]
Thank you @ataulm and @hitherejoe !! It is a great answer, so I would like to close this issue now.
